mindmeld/active_learning/alp.py

```diff
@@ -0,0 +1,273 @@
+            self.save,
+            self.training_level,
+            self.train_pattern,
+            self.test_pattern,
+            self.unlabeled_logs_path,
+            self.labeled_logs_pattern,
+            self.log_usage_pct,
+        )
+        logger.info("Starting Selection.")
+        self._train_single_strategy(
+            strategy=self.selection_strategy, training_mode=False
+        )
+        self.results_manager.write_log_selected_queries_json(
+            strategy=self.selection_strategy,
+            queries=self.data_bucket.newly_sampled_queries,
+        )
+
+    def plot(self):
+        """Creates the generated folder and its subfolders if they do not already exist."""
+        plot_manager = PlotManager(self.results_manager.experiment_folder)
+        plot_manager.generate_plots()
+
+    def _train_all_strategies(self):
+        """ Train with all active learning strategies."""
+        for strategy in self.training_strategies:
+            self._train_single_strategy(strategy)
+
+    def _train_single_strategy(self, strategy: str, training_mode: bool = True):
+        """Helper function to traing a single strategy.
+
+        Args:
+            strategy (str): Single strategy to train
+            training_mode (bool): If True, accuracies will be recorded. If False, accuracies
+                will not be recorded and run will terminate after first iteration.
+        """
+        heuristic = HeuristicsFactory.get_heuristic(strategy)
+        for epoch in range(self.n_epochs):
+            self._reset_data_bucket()
+            for iteration in range(self.num_iterations):
+                self._log_training_status(strategy, epoch, iteration)
+                # Run training and obtain probability distributions for each query
+                (
+                    eval_stats,
+                    confidences_2d,
+                    confidences_3d,
+                    confidence_segments,
+                ) = self.mindmeld_al_classifier.train(self.data_bucket, heuristic)
+
+                if training_mode:
+                    self._save_training_data(strategy, epoch, iteration, eval_stats)
+
+                num_unsampled = len(self.data_bucket.unsampled_queries)
+                if num_unsampled > 0:
+                    self.data_bucket.sample_and_update(
+                        sampling_size=self._get_sampling_size(num_unsampled),
+                        confidences_2d=confidences_2d,
+                        confidences_3d=confidences_3d,
+                        heuristic=heuristic,
+                        confidence_segments=confidence_segments,
+                    )
+                # Terminate on the first iteration if in selection mode.
+                if not training_mode:
+                    return
+
+    def _reset_data_bucket(self):
+        """ Reset the DataBucket to the initial DataBucket after every epoch."""
+        self.data_bucket = deepcopy(self.init_train_data_bucket)
```

Comment:
I think this is a very heavy operation, we need to store indices of the queries in order to do this

Reply:
Update: Saving the initial query ids for sampled and unsampled queries. Should be a lot less memory intensive now